### PR TITLE
s3hash: Add CRC64NVME checksum support

### PR DESCRIPTION
**Pull request by nl0**
## Summary
Refactor s3hash lambda to algorithm-agnostic MPU computation utility with explicit algorithm parameter.

### Algorithm-Agnostic Architecture
- Accept explicit `checksum_algorithm` parameter from S3HashLambdaParams
- Support both CRC64NVME and SHA256_CHUNKED via single code path
- Remove legacy SHA256 (non-chunked) streaming computation entirely
- Remove CHUNKED_CHECKSUMS environment variable

### MPU-based Computation
- Update `create_mpu()` to use configurable ChecksumAlgorithm
- Update `PartUploadResult` to extract appropriate checksum field (ChecksumCRC64NVME or ChecksumSHA256)
- Update `compute_part_checksums()` to handle any algorithm based on parameter

### CRC64NVME Composition
- Implement `combine_crc64nvme()` using reflected polynomial (0x9a6c9329ac4bc9b5)
- Handle single-part uploads (use checksum directly)
- Handle multi-part uploads (combine part checksums using proper CRC math)

### Simplified Orchestration
- Remove GetObjectAttributes logic (moved to pkgpush for priority-based selection)
- Remove automatic algorithm detection
- Use `empty_crc64nvme()` or `empty_sha256_chunked()` based on algorithm parameter

### Dependencies
- Update quilt-shared to 12a7a24e (includes checksum_algorithm parameter)

## Testing
✅ **Verified on CRC stack (https://crc.dev.quilttest.com)**
- Commit: 58064eb9
- Tested with 204 files (545 GB FASTQ dataset)
- CRC64NVME computation: ~0.2s average per file
- SHA256_CHUNKED: maintains backward compatibility
- Ready for production rollout

## Context
Simplified s3hash to pure MPU computation utility. Strategy selection (precomputed vs compute, priority order) moved to pkgpush lambda for better separation of concerns.

Depends on: #4623

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

Successfully refactors s3hash lambda to support CRC64NVME alongside SHA256_CHUNKED checksums through an algorithm-agnostic MPU computation architecture.

## Key Changes
- Removed legacy streaming SHA256 computation and `GetObjectAttributes` logic (moved to pkgpush)
- Added explicit `checksum_algorithm` parameter to `lambda_handler()` and `compute_checksum()`
- Updated `create_mpu()` to accept `ChecksumAlgorithm` and pass to S3 API
- Modified `PartUploadResult` to extract algorithm-specific checksum fields dynamically via `algorithm.s3_checksum_field`
- Implemented algorithm-specific checksum combination using `TX.assert_never()` for exhaustive checking
- Removed `CHUNKED_CHECKSUMS` environment variable

## Test Coverage
- Comprehensive parametrized tests for both CRC64NVME and SHA256_CHUNKED algorithms
- Tests cover empty files, single-part MPU, multi-part MPU, and copy operations
- Added edge case coverage for MPU abort failures and double completion
- All tests properly mock the `combine_crc64nvme()` function from py-shared

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge - the refactoring is well-architected with excellent test coverage and proper type safety
- Clean refactoring with explicit algorithm parameter passing, exhaustive type checking using TX.assert_never(), comprehensive test coverage for both algorithms, proper error handling, and verified in production-like environment (CRC stack with 204 files/545GB dataset)
- No files require special attention

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| lambdas/s3hash/src/t4_lambda_s3hash/__init__.py | 5/5 | Refactored to algorithm-agnostic MPU computation with explicit ChecksumAlgorithm parameter support for both CRC64NVME and SHA256_CHUNKED |
| lambdas/s3hash/tests/test_compute_checksum.py | 5/5 | Comprehensive test coverage for both CRC64NVME and SHA256_CHUNKED algorithms with proper mocking and edge case coverage |
| lambdas/s3hash/tests/test_wrapper.py | 5/5 | Added comprehensive MPU error handling tests including abort failure and double completion edge cases |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client as pkgpush Lambda
    participant Handler as lambda_handler
    participant Compute as compute_checksum
    participant S3API as S3 API
    participant MPU as create_mpu
    participant Upload as upload_parts
    participant Combine as Checksum Combination

    Note over Client,Combine: CRC64NVME Checksum Computation Flow

    Client->>Handler: Invoke with checksum_algorithm=CRC64NVME
    Handler->>Compute: compute_checksum(location, scratch_buckets, algorithm)
    
    Compute->>S3API: head_object(location)
    S3API-->>Compute: ETag, ContentLength
    
    alt Empty File
        Compute->>Combine: Checksum.get_empty(CRC64NVME)
        Combine-->>Compute: empty_crc64nvme()
        Compute-->>Handler: ChecksumResult
    else Non-Empty File
        Compute->>Compute: get_parts_for_size(total_size)
        Compute->>S3API: get_bucket_region(bucket)
        S3API-->>Compute: region
        Compute->>Compute: get_mpu_dst_for_location(scratch_buckets)
        
        Compute->>MPU: create_mpu(mpu_dst, CRC64NVME)
        MPU->>S3API: create_multipart_upload(ChecksumAlgorithm="CRC64NVME")
        S3API-->>MPU: UploadId
        
        Compute->>Upload: compute_part_checksums(mpu, location, etag, parts, algorithm)
        loop For each part
            Upload->>S3API: upload_part_copy(PartNumber, CopySourceRange)
            S3API-->>Upload: ChecksumCRC64NVME (base64), ETag
        end
        Upload-->>Compute: part_checksums (base64 decoded bytes)
        
        Compute->>MPU: __aexit__ (abort MPU)
        MPU->>S3API: abort_multipart_upload()
        
        alt CRC64NVME Algorithm
            Compute->>Combine: Checksum.crc64nvme_from_parts(checksums, part_sizes)
            Combine->>Combine: combine_crc64nvme(part_crcs, part_sizes)
            Combine-->>Compute: Combined CRC64NVME checksum
        else SHA256_CHUNKED Algorithm
            Compute->>Combine: Checksum.sha256_chunked_from_parts(checksums)
            Combine-->>Compute: Double-hashed SHA256 checksum
        end
        
        Compute-->>Handler: ChecksumResult(checksum)
    end
    
    Handler-->>Client: ChecksumResult
```

### CODE REVIEW

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py [hunk 1]
```diff
@@ -167,10 +167,30 @@ def get_compliant_checksum(attrs: GetObjectAttributesOutputTypeDef) -> T.Optiona
     return None
 
 
+def get_aws_crc64nvme_checksum(attrs: GetObjectAttributesOutputTypeDef) -> T.Optional[Checksum]:
+    """Retrieve precomputed CRC64NVME checksum from S3 object attributes (Tier 1)."""
+    checksum_value = attrs.get("Checksum", {}).get("ChecksumCRC64NVME")
+    if checksum_value is None or attrs["ObjectSize"] == 0:
+        logger.info(f"[PERF] get_aws_crc64nvme_checksum: No precomputed CRC64NVME (size={attrs.get('ObjectSize', 0)})")
```
Comment (sir-sigurd) [inline lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: doesn't make sense to do `.get('ObjectSize', 0)`

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py [hunk 2]
```diff
@@ -372,25 +394,76 @@ async def compute_checksum_legacy(location: S3ObjectSource) -> Checksum:
     return Checksum.sha256(hashobj.digest())
 
 
+def combine_crc64nvme(part_crcs: T.List[bytes], part_sizes: T.List[int]) -> bytes:
```
Comment (sir-sigurd) [inline lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]:
```suggestion
def combine_crc64nvme(part_crcs: list[bytes], part_sizes: list[int]) -> bytes:
```

Comment (sir-sigurd) [inline lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: probably we could complete MPU to make it calculate it for us
but I think I had impression I saw it could be relatively slow

from other side code for combining CRC could be useful anyway

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py [hunk 3]
```diff
@@ -403,7 +476,19 @@ async def compute_checksum(location: S3ObjectSource, scratch_buckets: T.Dict[str
             part_defs,
         )
 
-    checksum = Checksum.for_parts(part_checksums)
+    # Combine per-part CRC64NVME checksums into whole-file checksum
+    if len(part_checksums) == 1:
```
Comment (sir-sigurd) [inline lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: do we need a special case here?

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py (hunk 2 above)
Comment (nl0) [reply lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: yeah i think it's faster to just combine them, since we already have all that data.
plus, we don't have to clean up after ourselves in this case, since the unfinished mpus will be garbage-collected.

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py (hunk 3 above)
Comment (nl0) [reply lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: yeah we probably don't

File: lambdas/s3hash/src/t4_lambda_s3hash/__init__.py [hunk 4]
```diff
@@ -372,25 +394,76 @@ async def compute_checksum_legacy(location: S3ObjectSource) -> Checksum:
...
+
 async def compute_checksum(location: S3ObjectSource, scratch_buckets: T.Dict[str, str]) -> ChecksumResult:
+    logger.info(f"[PERF] compute_checksum START: {location}")
+
+    # Tier 1: Try to get precomputed CRC64NVME from S3
+    logger.info(f"[PERF] Tier 1: Attempting GetObjectAttributes for {location}")
     obj_attrs = await get_obj_attributes(location)
     if obj_attrs:
-        checksum = get_compliant_checksum(obj_attrs)
+        checksum = get_aws_crc64nvme_checksum(obj_attrs)
```
Comment (sir-sigurd) [inline lambdas/s3hash/src/t4_lambda_s3hash/__init__.py]: I think we still need to use chunked sha256 when it's available because files uploaded with quilt3 don't have CRC

File: lambdas/s3hash/tests/test_compute_checksum.py [hunk 5]
```diff
@@ -7,7 +9,28 @@
 from pytest_mock import MockerFixture
 
 import t4_lambda_s3hash as s3hash
-from quilt_shared.const import MAX_PART_SIZE
+from quilt3.data_transfer import CHECKSUM_MULTIPART_THRESHOLD
+from quilt_shared.crc64 import combine_crc64nvme
+
+
+def test_combine_crc64nvme():
```
Comment (sir-sigurd) [inline lambdas/s3hash/tests/test_compute_checksum.py:16]: why is it tested here?

File: lambdas/s3hash/tests/test_wrapper.py [hunk 6]
```diff
@@ -95,6 +115,81 @@ def dict(self):
...
     )
 
     aio_context_mock.assert_called_once_with(s3hash.AWSCredentials.parse_obj(AWS_CREDENTIALS))
+
+
+@pytest.mark.asyncio
+async def test_mpu_abort_failure(s3_stub: Stubber, mocker: MockerFixture):
+    """Test that MPU abort errors don't crash the cleanup.
+
+    This tests the error recovery path in create_mpu context manager (lines 279-283)
```
Comment (sir-sigurd) [inline lambdas/s3hash/tests/test_wrapper.py:128]: don't mention code lines

File: lambdas/s3hash/tests/test_wrapper.py [hunk 7]
```diff
@@ -95,6 +115,81 @@ def dict(self):
...
+
+    # Verify the exception was logged
+    mock_logger.exception.assert_called_once_with("Error aborting MPU")
+
+
+@pytest.mark.asyncio
+async def test_mpu_double_completion(s3_stub: Stubber):
+    """Test that completing MPU twice raises an error.
+
+    This tests the defensive check in MPURef.complete() (lines 246-248)
```
Comment (sir-sigurd) [inline lambdas/s3hash/tests/test_wrapper.py:175]: don't mention code lines

**Review by sir-sigurd — APPROVED** [review]

File: lambdas/s3hash/tests/test_compute_checksum.py (hunk 5 above)
Comment (nl0) [reply lambdas/s3hash/tests/test_compute_checksum.py:16]: it's a leftover -- the combination code was here initially
